### DISCUSSION TITLE
Is there an idiomatic way to return a "success-or-error" value with nullability checks?

### DISCUSSION THREAD

**Discussion by anthony-c-martin** [dotnet/roslyn#56962] [Q&A] [answered]
I've been using nullability enabled and pattern matching syntaxes heavily, and often find myself wanting to write a lot of code like this:

```csharp
(TSuccess? success, TError? error) TryDoSomething()
{
  ...
  if (someCond)
  {
    return (success, null);
  }
  else
  {
    return (null, error);
  }
}
```

When handling the result of `TryDoSomething()`, I know that exactly one of `error` or `success` will always be non-null. Is there a way to signal this to the compiler?

Without this, I'm then forced to write pattern matching code that looks like:
```csharp
TryDoSomething() switch {
  (success, null) => ...,
  (null, error) => ...,
  _ => throw new InvalidOperationException("This will never happen, but I have to handle it anyway"),
}
```

In languages with union types, this pattern is pretty common (e.g. Typescript):
```typescript
function tryDoSomething(): { error: Error } | { success: Success } {
  ...
}
```

Rust's enum types provide a really neat solution:
https://doc.rust-lang.org/std/result/

I'm curious if there's an established pattern to follow here, or if people have come up with creative solutions to it.

**Comment by PathogenDavid** [comment, marked as answer]
> I'm curious if there's an established pattern to follow here

Not really. The typical pattern in the .NET ecosystem has always been either exceptions or the detail-barren `bool TryWhatever(out TResult result)` pattern.

> or if people have come up with creative solutions to it.

I don't think there's anything quite as elegant as what you're used to with TypeScript or Rust until C# gets proper support for union types.

You could use an analyzer to automatically suppress CS8509 and avoid the catch-all `_` leg of your switch statement. (You might want to use something other than a tuple though to identify your unions.)

You'll likely find this work-in-progress solution to be interesting: https://github.com/dotnet/csharplang/discussions/5244

--------------

Proper union type support in C# is tracked by the following discussions:

* https://github.com/dotnet/csharplang/issues/113
* https://github.com/dotnet/csharplang/discussions/3760

They were briefly considered for C# 10, but it was decided there wasn't enough time: [April 7th, 2021 meeting notes](https://github.com/dotnet/csharplang/blob/6e57bef9536d1ff08dd4fb99f6388b4b233e1b0d/meetings/2021/LDM-2021-04-07.md#discriminated-unions)

The language designers have expressed interest in giving them attention for C# 11: [August 30th, 2021 meeting notes](https://github.com/dotnet/csharplang/blob/6e57bef9536d1ff08dd4fb99f6388b4b233e1b0d/meetings/2021/LDM-2021-08-30.md#discriminated-unions)

**Reply by anthony-c-martin** [reply]
Thanks for the detailed response!

**Comment by jnm2** [comment]
I do variations on a hand-rolled discriminated union, a struct `Result<T>`: https://gist.github.com/jnm2/a0031a31d04d6ed4d44d8dfc9265b8cf.

`Result.Success(value)` and `Result.Error("message")` to produce, and `result.IsError(out var message)` or `result.IsSuccess(out var value)` or `result.Value` (possibly throwing InvalidOperationException just like `Nullable<T>`) to consume.
